src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala

```diff
@@ -124,14 +130,21 @@ abstract class Pickler extends SubComponent {
     override protected def shouldSkipThisPhaseForJava: Boolean = !settings.YpickleJava.value
   }
 
+  // temporary variables used to populate a pickle
+  // reuse is handled by Pickle#active: each pickle gets to use them until `finish` is called,
+  // after which access will throw an assertion error.
+  private var index   = new mutable.AnyRefMap[AnyRef, Int]
+  private var entries = new Array[AnyRef](256)
+
   private class Pickle(root: Symbol) extends PickleBuffer(new Array[Byte](4096), -1, 0) {
     private val rootName  = root.name.toTermName
     private val rootOwner = root.owner
-    private var entries   = new Array[AnyRef](256)
     private var ep        = 0
-    private val index     = new mutable.AnyRefMap[AnyRef, Int]
+    private var active    = true
```

Comment:
It would be more obvious to me if `class Pickle(root, var index, var entries)` and `finish = { index = null; entries = null }` and managing the outer index, entries was handled in outer class.

Then `Using.Releasable[Pickle]` where `release(p) = p.finish()` and `Using(new Pickle(...))`.

Reply:
So this is a 2.12 PR because that's what I use, passé though that may be. I tarted it up with a similar implementation; someone with a real thing for type safety would have made an `abstract class ImmutablePickle extends PickleBuffer` so you couldn't call the methods that may crash, but I'm not that cool.